airflow/providers/google/cloud/hooks/compute_ssh.py

```diff
@@ -0,0 +1,378 @@
+            pubkey = f"{pkey_obj.get_name()} {pkey_obj.get_base64()} {username}"
+        except (IOError, SSHException) as err:
+            raise AirflowException(f"Error encountered creating ssh keys, {err}")
+
+        if use_oslogin:
+            # Expiration time is in microseconds.
+            self.set_oslogin_key(pubkey, expire_time)
+        else:
+            # using instance metadata to manage keys
+            self.set_sshkey_metadata(project_id=project_id,
+                                     zone=zone,
+                                     instance=instance,
+                                     username=username,
+                                     pubkey=pubkey)
+        return pkey_obj
+
+    @GoogleBaseHook.fallback_to_default_project_id
+    def run(self,
+            cmd: str, instance: str, zone: str,
+            project_id: Optional[str] = None,
+            expire_time: int = 300,
+            hostname: Optional[str] = None,
+            use_internal_ip: bool = False,
+            use_iap_tunnel: bool = False,
+            use_oslogin: bool = True
+            ) -> Any:
+        """
+         Execute a command in a remote instance
+
+        :param cmd: The command to execute on the instance
+        :type cmd: str
+        :param instance: The name of the instance the command will be executed on
+        :type instance: str
+        :param zone: The zone where the instance is located
+        :type zone: str
+        :param project_id: The project ID of the instance
+        :type project_id: Optional[str]
+        :param expire_time: The maximum amount of time before the private key is expired
+        :type expire_time: int
+        :param hostname: The hostname of the instance.
+        :type hostname: str
+        :param use_iap_tunnel: Whether to connect through IAP tunnel
+        :type use_iap_tunnel: bool
+        :param use_internal_ip: Whether to connect using internal IP
+        :type use_internal_ip: bool
+        :param use_oslogin: Whether to manage keys using OsLogin API. If false,
+            keys are managed using instance metadata
+        """
+        hostname = self.get_connection_type(project_id=project_id,
+                                            instance=instance,
+                                            zone=zone,
+                                            hostname=hostname,
+                                            use_iap_tunnel=use_iap_tunnel,
+                                            use_internal_ip=use_internal_ip
+                                            )
+
+        username = self.get_username(oslogin=use_oslogin)
+
+        self.log.info("Creating remote connection to host")
+        pkey = self._create_ssh_key(
+            instance=instance,
+            project_id=project_id,
+            zone=zone,
+            expire_time=expire_time,
+            use_oslogin=use_oslogin
+        )
+        sshclient = self.get_sshclient(pkey=pkey,
```

Comment:
There is one problem if internal IP addresses are used.  Two servers can use the same IP address if they are on two different networks. We should make sure that we have connected to the expected server.
```
metadata_id_url = 'http://metadata.google.internal/computeMetadata/v1/instance/id'
remote_command = f'[ `curl "{metadata_id_url}" -H "Metadata-Flavor: Google" -q` = {instance_id} ] || exit 23'
```